### ISSUE TITLE
Add new labels to manage ATDM Trilinos GitHub Issues

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#3963]
CC List:
* @trilinos/framework 
* @maherou (Trilinos overall Lead)
* @jwillenbring (Trilinos Framework Product Area Lead)
* @kddevin  (Trilinos Data Services Product Area Lead)
* @srajama1 (Trilinos Linear Solvers Product Area Lead)
* @rppawlo  (Trilinos Nonlinear Solvers Product Area Lead)
* @mperego (Trilinos Discretizations Product Area Lead)
* @fryeguy52 (ATDM Tools Member)

## Description

This Issue is to document the need for and to add some new labels to improve the management of ATDM Trilinos GitHub issues as part of the ATDM Trilinos triaging and issue management process currently documented at:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/Triaging+and+addressing+ATDM+Trilinos+Failures

There are a few problems with that process as currently documented:

* **Problem-1:** There is currently no way to quickly identify how serious a failure on the CDash Dashboard is ATDM Trilinos is w.r.t. the ATDM APPs.  Does this failure really bad for the APP or does it not impact that APP usage of Trilinos at all?  (We can say it in words in the GitHub Issue text but it is hard to query based on that and hard to find in the issue text.)

* **Problem-2:** There is currently no convenient way to find all of the ATDM Trilinos GitHub issues that belong to a given Trilinos Product Area (for which each Trilinos Product Area Lead has the responsibility to oversee).  (One can search based on the package labels but GitHub does not provide a way to `OR` a bunch of labels in an Issue query and such a query would be very long and hard to maintain even if GitHub supported this.)

* **Problem-3:** Some Trilinos failures on ATDM platforms are due to problems with the env and not really Trilinos code or test issues.  It would be useful to segragate those issues that are likely Trilinos code/test bugs and thosse are are casued by defects in the env outside of Trilins.

## Proposed Solution

For **Problem-1** the proposed solution is to define three new Trilinos GitHub labels to represent the severity of the problem as it relates to that ATDM APPs usage of Trilinos:

* `ATDM Critical`: Problems that critically damage the ability to even run the ATDM Trilinos builds.
  - Examples:
    - A library build failure on an important platform (e.g. CUDA on 'waterman') that takes out hundreds of tests.
    - A runtime defect in an upstream package that takes out hundreds of
  - Related practices:
    - The issue is fixed with STOP THE LINE urgency:
      - The problem is fixed ASAP, or
      - The PR that introduced the bug is reverted
    - ATDM APPs will not get an update of Trilinos until this problem is resolved
    - Failures in any of the associated tests result a global FAILED status for that ATDM Trilinos status CDash analysis tool (see #2933)

* `ATDM Blocker`: Problems that make Trilinos unfit to be adopted by one or more ATDM APPs but do not seriously damage ability to run automated builds
  - Examples:
    - A runtime issue that breaks the functioning of an important Trilinos capability used by an ATDM APP but only impacts a small number of Trilinos tests
  - Related practices:
    - The issue does not need to be resolved with STOP THE LINE urgency but needs to be fixed before the next Trilinos update for the affected APP.  (For example, a problem with Phalanx might block EMPIRE getting an updated version of Trilinos but should not block SPARC from getting an updated version of Trilinos.)
    - The impacted APP should not get an updated version of Trilinos until the issue is resolved
    - Failures in any of the associated tests result a global FAILED status for that ATDM Trilinos status CDash analysis tool (see #2933)

* `ATDM Nonblocker`: Problems with the Trilinos that don't impact ATDM APPs
  - Examples:
    - Failures in some tests for solvers that are not used by any of the ATDM APPs
    - A defect in a Trilinos test on a small number of platforms that has been confirmed not to be a bug in Trilinos library code
  - Related practices:
    - The associated tests are marked as `okay_to_fail=1` in the tests with issue tracker files for the ATDM Trilinos CDash analysis tool (see #2933), or
    - The associated code and/or tests are disabled in ATDM Trilinos builds going forward
    - Failures in any of the non-disabled associated tests will **NOT** result a global FAILED status for that ATDM Trilinos status CDash analysis tool (see #2933)  (That is, it will be allowed to fail but we will still be able to track if it passes or fails.)

For **Problem-2** the proposed labels for the [Trilinos Product Areas](https://snl-wiki.sandia.gov/display/TRIL/Trilinos+Product+Areas+and+Product+Leaders) are:

* `PA: Framework`: All GitHub issues that fall under the Trilinos Framework Product Area

* `PA: Data Services`: All GitHub issues that fall under the Trilinos Data Services Product Area

* `PA: Linear Solvers`: All GitHub issues that fall under the Trilinos Linear Solvers Product Area

* `PA: Nonlinear Solvers`: All GitHub issues that fall under the Trilinos Nonlinear Linear Solvers Product Area

* `PA: Discretizations`: All GitHub issues that fall under the Trilinos Discretizations Product Area

For **Problem-3**, there is the label:

* `ATDM Env Issue`: Problem mostly caused by the env and not a defect in Trilinos code or tests
  - Examples:
    - A bug in an LAPACK function or in calling that function (e.g. #2410, #3497)
    - A bug in MPI (e.g. #3331, #3290)
  - Related practices:
    - Provide separate statistics for ATDM Trilinos issues based on those with the label `ATDM Env Issue` and those without it (i.e. are actual Trilinos code or test issues)
    - Would not generally block the update of Trilinos for an ATDM APP unless changes in Trilinos exposed the env problem but the old Trilinos version avoided it.   (In that latter case, ATDM APPs should not get an update of Trilinos since the current version of Trilinos they are using does not trigger this problem with the env).

**Comment by bartlettroscoe** [comment]
FYI: I added the following Confluence page to document these new labels:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels

**Comment by bartlettroscoe** [comment]
@maherou (Trilinos overall Lead)
@jwillenbring (Trilinos Framework Product Area Lead)
@kddevin (Trilinos Data Services Product Area Lead)
@srajama1 (Trilinos Linear Solvers Product Area Lead)
@rppawlo (Trilinos Nonlinear Solvers Product Area Lead)
@mperego (Trilinos Discretizations Product Area Lead)

Any time to look over the list of labels used with the ATDM Trilinos GitHub issues (and their meaning) defined at:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels

@srajama1, at the ATDM/Trilinos Testing Meeting last Friday, you commented that the ATDM Trilinos GitHub issues have too many labels.  Can you please comment on which labels listed at the site that you would get rid of and why?

**Comment by mperego** [comment]
@bartlettroscoe I cannot access that link. If the link is correct than I don't have the right permissions.

**Comment by mperego** [comment]
@bartlettroscoe Thanks, it works now!

**Comment by srajama1** [comment]
If we have 3 labels per issue criticality, product area, bug/enhancement/question when it gets created, I am fine with it. We could package name and other labels later. Why do you need Framework and PA: Framework ? I would also get rid of ATDM as criticality has ATDM as a prefix anyway.

**Comment by bartlettroscoe** [comment]
> If we have 3 labels per issue criticality, product area, bug/enhancement/question when it gets created, I am fine with it. We could package name and other labels later. Why do you need Framework and PA: Framework ? I would also get rid of ATDM as criticality has ATDM as a prefix anyway.

@srajama1, if GitHub Issue queries supported ORing labels, I would agree with you.  The problem is that if you remove the "ATDM" label then there is no way to query all of the issues impacting ATDM.  (Also, "enhancement" and "question" issues don't have a criticality label.)

Basically, all of these labels are needed to provide the GitHub issue views provided at:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM+Builds+of+Trilinos#ATDMBuildsofTrilinos-QuickLinks

(Even if you don't need to see all of these views, we do.  So to remove some of these labels makes our job harder.)

If you get rid of the package label, then there is no way to see, for example, all of the Anasazi GitHub issues impacting ATDM (which is important to see in cases like [TRIL-238](https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-238)):

* https://github.com/trilinos/Trilinos/issues?utf8=%E2%9C%93&q=is%3Aopen+label%3AATDM+label%3AAnasazi+

(And we are going to need to do queries and create stat reports by package to see which packages are most problematic for ATDM.  That will be needed to inform staff and management about where resources need to go to improve things.)

If we get rid of the "Disabled Tests" label, then it is not possible to filter out GitHub issues that have been (partially) resolved by simply disabling tests.  (And people have asked us **not** to close issues that have been resolved by disabling tests.)  But the label "Disabled Tests" label should never show up in the GitHub issues or Issue queries that you are dealing with (because we filter those out of your view).  

The labels "ATDM Config" and "ATDM DevOps" should likely never show up in the GitHub issues or GitHub Issue queries that you are having to address.

I agree that the "Framework" and "PA: Framework" labels may be redundant.  But those labels should never show up in the GitHub Issue queries or Issues that you are responsible for.  So it is really up to the Trilinos Framework Product Area Lead (i.e. @jwillenbring) to make the call on that.  While the purpose of the label "PA: Framework" is well defined, the meaning of the label "Framework" is not.  (But that is @jwillenbring's problems, not your).

One label that seems like we should not need it is the "Stalled" label.  There are very few issues that are given that label as you can see in:

* https://github.com/trilinos/Trilinos/issues?utf8=%E2%9C%93&q=is%3Aopen+label%3Astalled+-label%3A%22Disabled+Tests%22+

These are basically used for non-critical issues that are just not getting worked on or resolved.  We don't want to close the issue but we don't want them to clutter up the issues that don need to get worked on.    But, again, you will not see GitHub issues in your queries because we filter them out.

How do other people feel about the usage of GitHub Issue Labels?

**Comment by bartlettroscoe** [comment]
Also, I think we need the "ATDM Env Issue" label to be able to provide statistics about ATDM Trilinos GitHub issues that are caused by problems/changes in the env vs. Trilinos bugs.  Otherwise, it does not seem fair to ding the Trilinos Framework Product Area, for example, for an issue like #2410 that has been open and unresolved for nearly 9 months (which is caused by problems with LAPACK on the Power systems, not a problem with Trilinos as far as we can tell).  So we need to show separate stats for ATDM Trilinos GitHub issues **with** the label "ATDM Env Issue" (which shows how good the sysadmins are at addressing env issues) and show stats for Issues **without** the label "ATDM Dev Env" (which are bugs in Trilinos, which the Trilinos Product Area Leads and Trilinos developers **are** responsible to resolve).

So, in conclusion, given the views that we need to see, the stats we need to be able to compute, and the limitation of GitHub Issue queries, the only labels I think we could consider eliminating are:

* "Framework" (could be redundant with "PA: Framework" but @jwillenbring will need to decide that)
* "Stalled" (but then we will be forced to deal with these issues in some other way)

Anyone else agree or disagree with this?

**Comment by bartlettroscoe** [comment]
Since there have been no objections, I am closing this issue as complete.  The new labels are in usage.  (Now we just need to get people to assign the correct [ATDM Severity/Criticality Labels](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels#ATDM-SpecificTrilinosGitHubLabels-LabelsforSeverity/Criticality) and we should be good to go.)

**Comment by bartlettroscoe** [comment]
@jwillenbring (Trilinos Framework Product Area Lead)
@kddevin (Trilinos Data Services Product Area Lead)
@srajama1 (Trilinos Linear Solvers Product Area Lead)
@rppawlo (Trilinos Nonlinear Solvers Product Area Lead)
@mperego (Trilinos Discretizations Product Area Lead)

Given that there are a few GitHub labels that start with "ATDM" I think we should change the names of the [ATDM Severity/Criticality Labels](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels#ATDM-SpecificTrilinosGitHubLabels-LabelsforSeverity/Criticality) to `ATDM Sev: <level>`.  That will make them sort together and make them easier to find when adding labels in GitHub.

Any objections or suggestions for a better prefix for these labels?

P.S. I forgot to close the issue in my last comment so I will leave it open until we make this last change.

**Comment by bartlettroscoe** [comment]
> Any objections or suggestions for a better prefix for these labels?

Since there were no objections or suggestions, I updated the [ATDM GitHub Severity/Criticality](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM-Specific+Trilinos+GitHub+Labels#ATDM-SpecificTrilinosGitHubLabels-LabelsforSeverity/Criticality) labels to be `ATDM Sev: <severity>` and updated all documentation that I could find and and linked to [ATDM Builds of Trilinos Wiki](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM+Builds+of+Trilinos).

Closing this as complete.

**Comment by bartlettroscoe** [comment]
Closing as complete ...
